Add a mechanism to remember the last IP address of the response for the main resource request

whsieh opened the pull request:
#### 07c42006c3448110ecf079e2392b4fba15267ef0
<pre>
Add a mechanism to remember the last IP address of the response for the main resource request
<a href="https://bugs.webkit.org/show_bug.cgi?id=246128">https://bugs.webkit.org/show_bug.cgi?id=246128</a>
rdar://100831206

Reviewed by John Wilander.

Add a mechanism to keep track of the last IP address, corresponding to the main resource request&apos;s
response; this information is stashed in a side table on the network session, alongside the existing
map of first party host to CNAME.

* Source/WTF/wtf/PlatformHave.h:

Remove this feature flag, since the oldest macOS version we currently support (Big Sur) contains
this SPI.

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/DNS.cpp:
(WebCore::IPAddress::fromString):

Add a helper method to create an `IPAddress` from an IP address string (IPv4 or IPv6).

* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::IPAddress):

Add another explicit constructor that takes `WTF::HashTableEmptyValueType`, so that we can store
`IPAddress` as a value in a `HashMap`; see below.

(WTF::HashTraits&lt;WebCore::IPAddress&gt;::emptyValue):

Additionally add a `HashTraits` template class for `IPAddress`, which allows us to store
`WebCore::IPAddress` as a hash table value by implementing `emptyValue()`.

* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::networkSession const):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::resetParametersToDefaultValues):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setFirstPartyHostCNAMEDomain):
(WebKit::NetworkSession::firstPartyHostCNAMEDomain):
(WebKit::NetworkSession::resetFirstPartyDNSData):
(WebKit::NetworkSession::setFirstPartyHostIPAddress):
(WebKit::NetworkSession::firstPartyHostIPAddress):
(WebKit::NetworkSession::resetCNAMEDomainData): Deleted.

Rename `resetCNAMEDomainData` to `resetFirstPartyDNSData`, since we now use it to clear both first
party CNAME information as well as first party IP address information.

* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::lastRemoteIPAddress):

Add a helper function to grab the last resolved IP address of the network data task; this is
intended to be used after we get a response.

(WebKit::NetworkDataTaskCocoa::shouldApplyCookiePolicyForThirdPartyCNAMECloaking const):
(WebKit::NetworkDataTaskCocoa::updateFirstPartyInfoForSession):

Update the first party IP address, alongside the first party CNAME for the main resource.

(WebKit::NetworkDataTaskCocoa::applyCookiePolicyForThirdPartyCNAMECloaking):
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::hasCNAMEAndCookieTransformSPI): Deleted.

Remove this helper function, since we no longer need `-respondsToSelector` checks for
`_resolvedCNAMEChain` and `_cookieTransformCallback`.

Canonical link: <a href="https://commits.webkit.org/255507@main">https://commits.webkit.org/255507@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d7ad599cee8c643abeddb00eab02e71ec0c95bac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92711 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1925 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23297 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102427 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96713 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1925 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30276 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85098 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98575 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98374 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1298 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79203 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28230 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83212 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84069 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36681 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/79117 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34473 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18033 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27423 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3814 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38345 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40637 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/81739 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37201 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18484 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode (failure)") | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WTF/wtf/PlatformHave.h:
```diff
@@ -719,7 +719,6 @@
 
 #if PLATFORM(COCOA)
 #define HAVE_CF_PREFS_SET_READ_ONLY 1
-#define HAVE_CFNETWORK_CNAME_AND_COOKIE_TRANSFORM_SPI 1
```
I assume this is because the SPI is available on all supported platforms now.

johnwilander commented on Source/WebKit/NetworkProcess/NetworkSession.cpp:
```diff
@@ -323,38 +323,45 @@ void NetworkSession::setShouldEnbleSameSiteStrictEnforcement(WebCore::SameSiteSt
 }
 
-void NetworkSession::resetCNAMEDomainData()
+void NetworkSession::resetFirstPartyHostData()
```
I think host is too high level. Could this be resetFirstPartyDomainNameResolutionData()? Or resetFirstPartyDNSData()?

whsieh replied:
Both of those sound good! I'll go with `resetFirstPartyDNSData()`, since that's slightly more concise.

johnwilander commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -269,6 +271,7 @@ class NetworkSession : public CanMakeWeakPtr<NetworkSession> {
     WebCore::FirstPartyWebsiteDataRemovalMode m_firstPartyWebsiteDataRemovalMode { WebCore::FirstPartyWebsiteDataRemovalMode::AllButCookies };
     WebCore::RegistrableDomain m_standaloneApplicationDomain;
     HashMap<String, WebCore::RegistrableDomain> m_firstPartyHostCNAMEDomains;
+    HashMap<String, String> m_firstPartyHostIPAddresses;
```
Do we represent IP addresses as Strings elsewhere? We probably do. Just making sure there's no existing abstraction we should use.

whsieh replied:
Good catch! It looks like `WebCore::IPAddress` is already a thing, so we should definitely use that.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -66,6 +67,13 @@ void enableNetworkConnectionIntegrity(NSMutableURLRequest *) { }
 
+static NSString *lastRemoteIPAddress(NSURLSessionDataTask *task)
+{
+    // FIXME: In a future patch, this should adopt CFNetwork API that retrieves the original IP
```
Could you file a bug and link to it here, please?

whsieh replied:
Filed: https://bugs.webkit.org/show_bug.cgi?id=246428.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -140,15 +148,6 @@ static float toNSURLSessionTaskPriority(WebCore::ResourceLoadPriority priority)
 
-#if HAVE(CFNETWORK_CNAME_AND_COOKIE_TRANSFORM_SPI)
-// FIXME: Remove these selector checks when macOS Big Sur has shipped.
-// https://bugs.webkit.org/show_bug.cgi?id=215280
```
Should this bug be resolved now?

whsieh replied:
Yep — I'll resolve it as a dupe once this lands.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -163,17 +162,21 @@ static bool hasCNAMEAndCookieTransformSPI(NSURLSessionDataTask* task)
 {
-    if (!hasCNAMEAndCookieTransformSPI(m_task.get()) || !networkSession() || !networkSession()->networkStorageSession() || !networkSession()->networkStorageSession()->resourceLoadStatisticsEnabled() || requestURL.host().isEmpty())
+    auto* session = networkSession();
+    if (!session || !session->networkStorageSession() || !session->networkStorageSession()->resourceLoadStatisticsEnabled() || requestURL.host().isEmpty())
```
I wonder if the repeated use of this check for an active session and ITP should be broken out into a convenience function?

whsieh replied:
Yes! (I was planning to do that as a part of the next patch, though we can pull that into this instead). Will pull this out into a separate helper method, called `shouldApplyCookiePolicyForThirdPartyCNAMECloaking()`.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
```diff
@@ -163,17 +162,21 @@ static bool hasCNAMEAndCookieTransformSPI(NSURLSessionDataTask* task)
 void NetworkDataTaskCocoa::applyCookiePolicyForThirdPartyCNAMECloaking(const WebCore::ResourceRequest& request)
 {
-    if (!hasCNAMEAndCookieTransformSPI(m_task.get()) || isTopLevelNavigation() || !networkSession() || !networkSession()->networkStorageSession() || !networkSession()->networkStorageSession()->resourceLoadStatisticsEnabled())
+    if (isTopLevelNavigation() || !networkSession() || !networkSession()->networkStorageSession() || !networkSession()->networkStorageSession()->resourceLoadStatisticsEnabled())
```
Ditto.

johnwilander reviewed (changes requested):
OK to land but please see my comments. I especially think we should find a more detailed replacement name for resetCNAMEDomainData().

johnwilander commented on Source/WebCore/platform/network/DNS.cpp:
```diff
@@ -55,4 +59,17 @@ void stopResolveDNS(uint64_t identifier)
     WebCore::DNSResolveQueue::singleton().stopResolve(identifier);
 }
 
+std::optional<IPAddress> IPAddress::fromString(const String& string)
```
Nice!

johnwilander commented on Source/WebKit/NetworkProcess/NetworkSession.h:
```diff
@@ -269,6 +272,7 @@ class NetworkSession : public CanMakeWeakPtr<NetworkSession> {
     WebCore::FirstPartyWebsiteDataRemovalMode m_firstPartyWebsiteDataRemovalMode { WebCore::FirstPartyWebsiteDataRemovalMode::AllButCookies };
     WebCore::RegistrableDomain m_standaloneApplicationDomain;
     HashMap<String, WebCore::RegistrableDomain> m_firstPartyHostCNAMEDomains;
+    HashMap<String, WebCore::IPAddress> m_firstPartyHostIPAddresses;
```
Great to use IPAddress.

whsieh committed:
Add a mechanism to remember the last IP address of the response for the main resource request
https://bugs.webkit.org/show_bug.cgi?id=246128
rdar://100831206

Reviewed by John Wilander.

Add a mechanism to keep track of the last IP address, corresponding to the main resource request's
response; this information is stashed in a side table on the network session, alongside the existing
map of first party host to CNAME.

* Source/WTF/wtf/PlatformHave.h:

Remove this feature flag, since the oldest macOS version we currently support (Big Sur) contains
this SPI.

* Source/WebCore/PAL/pal/spi/cf/CFNetworkSPI.h:
* Source/WebCore/platform/network/DNS.cpp:
(WebCore::IPAddress::fromString):

Add a helper method to create an `IPAddress` from an IP address string (IPv4 or IPv6).

* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::IPAddress):

Add another explicit constructor that takes `WTF::HashTableEmptyValueType`, so that we can store
`IPAddress` as a value in a `HashMap`; see below.

(WTF::HashTraits<WebCore::IPAddress>::emptyValue):

Additionally add a `HashTraits` template class for `IPAddress`, which allows us to store
`WebCore::IPAddress` as a hash table value by implementing `emptyValue()`.

* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::networkSession const):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::resetParametersToDefaultValues):
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::setFirstPartyHostCNAMEDomain):
(WebKit::NetworkSession::firstPartyHostCNAMEDomain):
(WebKit::NetworkSession::resetFirstPartyDNSData):
(WebKit::NetworkSession::setFirstPartyHostIPAddress):
(WebKit::NetworkSession::firstPartyHostIPAddress):
(WebKit::NetworkSession::resetCNAMEDomainData): Deleted.

Rename `resetCNAMEDomainData` to `resetFirstPartyDNSData`, since we now use it to clear both first
party CNAME information as well as first party IP address information.

* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::lastRemoteIPAddress):

Add a helper function to grab the last resolved IP address of the network data task; this is
intended to be used after we get a response.

(WebKit::NetworkDataTaskCocoa::shouldApplyCookiePolicyForThirdPartyCNAMECloaking const):
(WebKit::NetworkDataTaskCocoa::updateFirstPartyInfoForSession):

Update the first party IP address, alongside the first party CNAME for the main resource.

(WebKit::NetworkDataTaskCocoa::applyCookiePolicyForThirdPartyCNAMECloaking):
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::hasCNAMEAndCookieTransformSPI): Deleted.

Remove this helper function, since we no longer need `-respondsToSelector` checks for
`_resolvedCNAMEChain` and `_cookieTransformCallback`.

Canonical link: https://commits.webkit.org/255507@main
